File: Wff.Desktop/Utils/Ffmpeg.fs
//ffmpeg, pactl and wl-randr commands
module Utils.ffmpeg

open System.Text.Json

open System.Diagnostics

let audioDevices =
    let _process =
        Process.Start(
            ProcessStartInfo(
                FileName = "/bin/bash",
                Arguments = "-c \"pactl list sources | grep Name\"", //TODO: Don't use grep for this
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            )
        )

    let mutable audio_devices = [ "None" ]
    let mutable has_error = false

    _process.OutputDataReceived.Add(fun sender ->
        sender.Data
        |> Option.ofObj
        |> Option.iter (fun data ->
            if data.TrimStart().StartsWith "Name:" then
                let device = data.Split("Name:").[1].Trim()
                audio_devices <- audio_devices @ [ data.Split("Name:").[1] ]))

    match _process with
    | null -> audio_devices 
    | p ->
        p.Start() |> ignore
        p.BeginOutputReadLine()
        p.WaitForExit()
        audio_devices

type FfmpegCodecTarget =
    | Video
    | Audio

let codecs target =
    let codecFilter =
        match target with
        | FfmpegCodecTarget.Audio -> "A"
        | FfmpegCodecTarget.Video -> "V"

    let _process =
        Process.Start(
            ProcessStartInfo(
                FileName = "ffmpeg",
                Arguments = "-encoders",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            )
        )

    let mutable codecs = [ "Default" ]

    _process.OutputDataReceived.Add(fun sender ->
        sender.Data
        |> Option.ofObj
        |> Option.iter (fun data ->
            if
                data.TrimStart().StartsWith codecFilter
                && data.Split(" ", System.StringSplitOptions.RemoveEmptyEntries).Length > 2
            then
                let codec = data.Split(" ", System.StringSplitOptions.RemoveEmptyEntries).[1]

                if codec <> "=" then
                    codecs <- codecs @ [ codec ]))

    match _process with
    | null -> codecs
    | p ->
        p.Start() |> ignore
        p.BeginOutputReadLine()
        p.WaitForExit()
        codecs

//only the output name is needed
type RandrOutput = {name:string}

let outputs =
    let command = "wlr-randr"

    let _process =
        Process.Start(
            ProcessStartInfo(
                FileName = command,
                Arguments = "--json",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            )
        )

    let mutable outputs = []
    let lock = new System.Threading.ManualResetEventSlim false
    let buffer = new System.Text.StringBuilder()
    _process.OutputDataReceived.Add(fun sender ->
        sender.Data |> Option.ofObj |>Option.iter (fun data ->
            buffer.AppendLine data |> ignore
        )
    )
    _process.Exited.Add(fun _ ->
        let randrObj = JsonSerializer.Deserialize<List<RandrOutput>> (buffer.ToString())
        match randrObj with
            |null -> ()
            | o ->
                o |> Seq.iter (fun output ->
                    printfn "%s" output.name
                    outputs <- outputs @ [output.name]    
                )
        lock.Set() |> ignore
    )
    match _process with
        |null -> ["eDP-1"]
        | p ->
            p.EnableRaisingEvents <- true
            p.Start() |> ignore
            p.BeginOutputReadLine()
            p.WaitForExit()
            lock.Wait()
            outputs

File: Wff.Desktop/Utils/Ffmpeg.cs
namespace Wff.Utils;
using System.Diagnostics;
using System.Text.Json;
using System.Collections.Generic;
using System.Threading.Tasks;

public static class Ffmpeg
{
    public async static Task<List<string>> AudioDevicesAsync()
    {
        var process = new Process(
        );
        process.StartInfo =
            new ProcessStartInfo()
            {
                FileName = "/bin/bash",
                Arguments = "-c \"pactl list sources | grep Name\"",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };

        List<string> audio_devices = new() { "None" };
        if (process is null) return audio_devices;
        process.OutputDataReceived += (s, e) =>
        {
            if (e.Data is null) return;
            if (e.Data.TrimStart().StartsWith("Name:"))
            {
                var device = e.Data.Split("Name:")[1].Trim();
                audio_devices.Add(device);
            }
        };
        process.Start();
        process.BeginOutputReadLine();
        await process.WaitForExitAsync();
        return audio_devices;
    }
    public async static Task<List<string>> CodecsAsync(FfmpegCodecTarget target)
    {
        var codecFilter = target switch
        {
            FfmpegCodecTarget.Audio => "A",
            FfmpegCodecTarget.Video => "V",
            _ => "",
        };
        var process = new Process();
        process.StartInfo = new()
        {
            FileName = "ffmpeg",
            Arguments = "-encoders",
            UseShellExecute = false,
            CreateNoWindow = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true
        };
        List<string> codecs = new() { "Default" };
        process.OutputDataReceived += (s, e) =>
        {
            if (e.Data is null) return;
            if (e.Data.TrimStart().StartsWith(codecFilter) && e.Data.Split(" ", System.StringSplitOptions.RemoveEmptyEntries).Length > 2)
            {
                var codec = e.Data.Split(" ", System.StringSplitOptions.RemoveEmptyEntries)[1];
                if (codec != "=") codecs.Add(codec);
            }
        };
        if (process is null) return codecs;
        process.Start();
        process.BeginOutputReadLine();
        await process.WaitForExitAsync();
        return codecs;
    }
    public async static Task<List<string>> OutputsAsync()
    {
        var command = "wlr-randr";
        var process = new Process();
        process.StartInfo = new()
        {
            FileName = command,
            Arguments = "--json",
            UseShellExecute = false,
            CreateNoWindow = true,
            RedirectStandardOutput = true,
            RedirectStandardInput = true,
        };
        var outputs = new List<string>() { };
        var tlock = new System.Threading.ManualResetEventSlim(false);
        var buffer = new System.Text.StringBuilder();
        process.OutputDataReceived += (s, e) =>
        {
            if (e.Data is null) return;
            var data = e.Data;
            buffer.AppendLine(data);
        };
        process.Exited += (s, e) =>
        {
            var jsonDoc = JsonDocument.Parse(buffer.ToString());
            foreach (var element in jsonDoc.RootElement.EnumerateArray())
            {
                if (element.TryGetProperty("name", out var nameProperty))
                {
                    outputs.Add(nameProperty.GetString() ?? string.Empty);
                }
            }
            tlock.Set();
        };
        if (process is null) return outputs;
        process.EnableRaisingEvents = true;
        process.Start();
        process.BeginOutputReadLine();
        await process.WaitForExitAsync();
        tlock.Wait();
        return outputs;
    }
}

public enum FfmpegCodecTarget
{
    Video,
    Audio
};

public record RandrOutput
{
    public required string name { get; set; }
};


